### PR TITLE
Revert "use sqlalchemy 1.3.x for mautrix-telegram"

### PR DESCRIPTION
**Pull request by dotlambda**
Reverts NixOS/nixpkgs#127486

That has to be done using `packageOverrides`. And please avoid introducing `pythonPackages.sqlalchemy_1_3`.

See https://github.com/NixOS/nixpkgs/pull/126872 for how it should be done (for mautrix-signal in that case).

### CODE REVIEW

**Comment by Ma27** [comment]
This should not be merged as-is, otherwise we'd break mautrix-telegram again. Also we have several versioned packages already and can catch potential conflicts using `pythonCatchConflictsHook`, so one would find problems during the derivation's build.

I don't have a strong opinion about whether to instantiate a new python package-set or do overrides manually.

Also (slightly) related: #127599.

**Comment by dotlambda** [comment]
> This should not be merged as-is, otherwise we'd break mautrix-telegram again.

I added a commit fixing it again.

> Also we have several versioned packages already

That doesn't mean we should introduce more. Let's not make an antipattern that people shouldn't copy more common.

**Comment by petabyteboy** [comment]
I don't fully understand why this is the better approach, but I don't have a strong opinion on it either.

File: pkgs/servers/mautrix-telegram/default.nix
Comment (petabyteboy) [inline pkgs/servers/mautrix-telegram/default.nix:7]: Please also override "name", since it will otherwise use the old version in the derivation name.

**Review by ghost — CHANGES_REQUESTED** [review]

File: pkgs/servers/mautrix-telegram/default.nix
Comment (dotlambda) [reply pkgs/servers/mautrix-telegram/default.nix:7]: Oops, I forgot to use `overridePythonAttrs`.

**Comment by dotlambda** [comment]
> I don't fully understand why this is the better approach

Because we don't want packages in `pkgs/development/python-modules` to use non-default versions, which is encouraged by providing an attribute.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 127707` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mautrix-telegram</li>
  </ul>
</details>

File: pkgs/servers/mautrix-telegram/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/mautrix-telegram/default.nix:6]:
```suggestion
    packageOverrides = final: prev: {
      sqlalchemy = prev.sqlalchemy.overridePythonAttrs (oldAttrs: rec {
```

Comment (dotlambda) [reply pkgs/servers/mautrix-telegram/default.nix:6]: Come on, the other one is way more common.

**Comment by petabyteboy** [comment]
@dotlambda can you explain again why versioned attributes are bad? It seems like creating an own python package set in every place where an old version of sqlalchemy is necessary will create a lot of duplicated lines, since more packages will need to be downgraded to sqlalchemy 1.3.x again.

**Comment by alyssais** [comment]
@dotlambda isn't forcing every package that needs sqlalchemy 1.3.x to do its own seperate override going to be horrible for eval times?  I strongly disagree with versioned attributes for widely used, backwards-incompatible packages being an antipattern.

**Comment by dotlambda** [comment]
As said above, packages in `python3Packages` cannot use those versioned attributes. But that's what will unevitably happen if we introduce them. Considering the few lines this override takes, it's a much better trade-off.

**Comment by alyssais** [comment]
Why can't we revert those cases where `python3Packages` incorrectly use versioned attributes, instead of reverting things like this that are totally valid uses of them?  Either way, we evidently can't always catch these things before they make get merged, but by choosing to revert the versioned attributes entirely, rather than only improper uses of them, we're just making things needlessly harder (and evaluation needlessly more complex) for legitimate uses of non-default Python package verisons.

**Comment by jonringer** [comment]
Pinning within python-modules is highly discouraged. This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

The preference to handling this is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

**Comment by jonringer** [comment]
> @dotlambda isn't forcing every package that needs sqlalchemy 1.3.x to do its own seperate override going to be horrible for eval times? I strongly disagree with versioned attributes for widely used, backwards-incompatible packages being an antipattern.

For python packages, we just hold off updating packages which require a new version. Once a majority of the ecosystem switches, then we can move everything together. Unfortunately there's no isolation mechanisms within python to do this more gracefully

**Comment by Ma27** [comment]
> This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

That's why we have the `pythonCatchConflictsHook`, correct?

**Comment by jonringer** [comment]
> That's why we have the `pythonCatchConflictsHook`, correct?

That, and there's likely to be a partial overwrite if using `buildEnv` or `pythonX.withPackages`

**Comment by dotlambda** [comment]
> That's why we have the `pythonCatchConflictsHook`, correct?

Imagine hitting such a conflict when building `home-assistant`, with thousands of Python dependencies. That's impossible to trace back and solve and hence should be avoided by doing the override on a per-application basis.

**Comment by dotlambda** [comment]
Before this change, building `mautrix-telegram.alembic` fails with
```
pythonCatchConflictsPhase
Found duplicated packages in closure for dependency 'SQLAlchemy':
  SQLAlchemy 1.4.16 (/nix/store/j79ir7r6l67l212ffsissjggpp2hmxs6-python3.8-SQLAlchemy-1.4.16/lib/python3.8/site-packages)
  SQLAlchemy 1.3.23 (/nix/store/n9v3gbh414g59c7dcwjbv4qjl172hwb4-python3.8-SQLAlchemy-1.3.23/lib/python3.8/site-packages)

Package duplicates found in closure, see above. Usually this happens if two packages depend on different version of the same dependency.
builder for '/nix/store/gxn12ml89c4wfqzgny6yici6i4barisx-python3.8-alembic-1.5.8.drv' failed with exit code 1
```
You're welcome for fixing this properly and thank you for avoiding non-solutions in the future!

**Pull request merged**
